### ISSUE TITLE
False Positive for S4158 when using a delegate to add to the collection

### ISSUE THREAD

**Issue by davidmilligan** [SonarSource/sonar-dotnet#5544]
S4158 is incorrectly calling out that the collection is empty when a delegate to the add method is passed to another method
```C#
        public static void Main()
        {
            var list = new List<string>();
            AddSomeStuff(list.Add);
            list.ForEach(Console.WriteLine); //< - S4158 here, but the program actually prints hello world
        }

        private static void AddSomeStuff(Action<string> add)
        {
            add("hello");
            add("world");
        }
```
